webauthn-security-key/libs/auth.js

```diff
@@ -0,0 +1,586 @@
+/*
+ * @license
+ * Copyright 2019 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+const express = require('express');
+const router = express.Router();
+const crypto = require('crypto');
+const fido2 = require('@simplewebauthn/server');
+const base64url = require('base64url');
+const fs = require('fs');
+const low = require('lowdb');
+const { check, validationResult } = require('express-validator');
+
+if (!fs.existsSync('./.data')) {
+  fs.mkdirSync('./.data');
+}
+
+const FileSync = require('lowdb/adapters/FileSync');
+const adapter = new FileSync('.data/db.json');
+const db = low(adapter);
+
+router.use(express.json());
+
+const RP_NAME = 'webauthn-security-key-codelab';
+const FIDO_TIMEOUT = 30 * 1000 * 60;
+const VALID_TRANSPORTS = Object.freeze([
+  'usb',
+  'ble',
+  'nfc',
+  'internal',
+  'cable',
+]);
+
+const SINGLE_FACTOR = 'sfa';
+const TWO_FACTOR = '2fa';
+
+// Generic message, because an attacker should not be able to determine that a password was correct by looking at the error messages
+const GENERIC_AUTH_ERROR_MESSAGE =
+  'Username or password incorrect or credential not found or user verification failed';
+
+db.defaults({
+  users: [],
+}).write();
+
+// ----------------------------------------------------------------------------
+// Utils
+// ----------------------------------------------------------------------------
+
+function isPasswordCorrect(username, password) {
+  // Always true: in this demo the password isn't actually checked, for simplicity
+  return true;
+}
+
+function getAuthType(credentials) {
+  // If one or more credential are registered, two-factor is set up
+  // If no credential is registered, single factor
+  return credentials.length > 0 ? TWO_FACTOR : SINGLE_FACTOR;
+}
+
+function csrfCheck(req, res, next) {
+  if (req.header('X-Requested-With') != 'XMLHttpRequest') {
+    res.status(400).json({ error: 'Invalid access' });
+    return;
+  }
+  next();
+}
+
+function fullSessionCheck(req, res, next) {
+  if (req.session.name !== 'main') {
```

Comment:
@agektmr This check is based on the fact that if the `main` session has been started, it means authentication has been completed (see code flow). Does that seem OK?

Reply:
@agektmr's answer from F2F chat:
Yes